modules/monitoring/includes/downtime.php

```diff
@@ -0,0 +1,444 @@
+ */
+class Downtime {
+	/**
+	 * object RecurringDowntimePool_Model
+	 */
+	public $model;
+
+	/**
+	 * NinjaDateTime schedule start
+	 */
+	public $start;
+
+	/**
+	 * NinjaDateTime schedule end
+	 */
+	public $end;
+
+	/**
+	 * array
+	 */
+	public $cmd_mappings = array(
+		'hosts' => 'SCHEDULE_HOST_DOWNTIME',
+		'services' => 'SCHEDULE_SVC_DOWNTIME',
+		'hostgroups' => 'SCHEDULE_HOSTGROUP_HOST_DOWNTIME',
+		'servicegroups' => 'SCHEDULE_SERVICEGROUP_SVC_DOWNTIME'
+	);
+
+	/**
+	 * Provides helpers for working with scheduled downtime.
+	 *
+	 * Downtime constructor.
+	 * @param $model object (RecurringDowntimePool_Model)
+	 * @throws Exception
+	 */
+	function __construct($model) {
+		$this->model = $model;
+		$this->start = $this->get_scheduled_start();
+		$this->end = $this->get_scheduled_end();
+	}
+
+	/**
+	 * Returns number's ordinal
+	 *
+	 * @param $number int|mixed
+	 * @return string number's ordinal
+	 * @throws Exception UnexpectedValueException
+	 */
+	public function get_ordinal($number) {
+		$number -= 1; // number 1 becomes idx 0 of $ordinals.
+		$ordinals = array('first', 'second', 'third', 'fourth');
+		if(!array_key_exists($number, $ordinals)) {
+			throw new UnexpectedValueException(
+				sprintf('Missing number ordinal map for idx %s', $number)
+			);
+		}
+		return $ordinals[$number];
+	}
+
+	/**
+	 * Returns a formatted downtime command.
+	 *
+	 * This method mainly exists to simplify testing.
+	 *
+	 * @param $obj_name string
+	 * @param $comment_prefix string comment prefix
+	 * @return array command map
+	 */
+	public function get_command_mappings($obj_name, $comment_prefix = '') {
+		$downtime_type = $this->model->get_downtime_type();
+		if(!array_key_exists($downtime_type, $this->cmd_mappings)) {
+			throw new UnexpectedValueException("Missing mapping for downtime type: $downtime_type");
+		}
+
+		return array(
+			'cmd' => $this->cmd_mappings[$downtime_type],
+			'obj_name' => $obj_name,
+			'start' => $this->start->getTimestamp(),
+			'end' => $this->end->getTimestamp(),
+			'is_fixed' => $this->model->get_fixed(),
+			'duration' => $this->model->get_duration(),
+			'author' => $this->model->get_author(),
+			'comment' => $comment_prefix . $this->model->get_comment()
+		);
+	}
+
+	/**
+	 * Converts command mappings to a Nagios-interpretable string format.
+	 *
+	 * @param $obj_name string
+	 * @param $comment_prefix string comment prefix
+	 * @return string Nagios external command
+	 */
+	public function get_command($obj_name, $comment_prefix = 'AUTO: ') {
+		$command = $this->get_command_mappings($obj_name, $comment_prefix);
+		$cmd_fmt = 'cmd;obj_name;start;end;is_fixed;0;duration;author;comment';
+		return str_replace(array_keys($command), array_values($command), $cmd_fmt);
+	}
+
+	/**
+	 * Converts day of week number to string, starting with monday.
+	 *
+	 * @param $number int
+	 * @return string day of week name
+	 * @throws Exception UnexpectedValueException
+	 */
+	public function dow_to_string($number) {
```

Comment:
How about using `[jddayofweek](https://www.php.net/manual/en/function.date.php)($number,1)` instead?

Reply:
Yes, absolutely. I had no idea about this function.

Reply:
According to the official PHP documentation for `jddayofweek`, the first day of the week is on Sunday. When testing in my env, it's on Monday. This worries me a bit and leads me to think that it's dynamic and uses the locale to figure out which standard to use. If this is the case, we might see intermittent issues using it, as mappings in the frontend are static (days 1-7, with start on Monday).

Reply:
Yes, probably a good point. A similar way using date could be:

`date('l', strtotime("Sunday +1 days\));`

Keeping the current implementation is OK for me, too.